nexus/src/app/saga.rs

```diff
@@ -165,66 +336,68 @@ impl super::Nexus {
                     undo_node, undo_error
                 ));
 
-                error!(log, "saga stuck";
-                    "saga_id" => runnable_saga.id.to_string(),
+                // TODO this log message does not belong here because if the
+                // caller isn't checking this then we won't log it.  We should
+                // probably make Steno log this since there may be no place in
+                // Nexus that's waiting for a given saga to finish.
+                error!(self.log, "saga stuck";
+                    "saga_id" => self.id.to_string(),
                     "error" => #%error,
                 );
             }
 
             error
         })
     }
+}
 
-    /// Starts the supplied `runnable_saga` and, if that succeeded, awaits its
-    /// completion and returns the raw `SagaResult`.
-    ///
-    /// This is a test-only routine meant for use in tests that need to examine
-    /// the details of a saga's final state (e.g., examining the exact point at
-    /// which it failed). Non-test callers should use `run_saga` instead (it
-    /// logs messages on error conditions and has a standard mechanism for
-    /// converting saga errors to generic Omicron errors).
-    #[cfg(test)]
-    pub(crate) async fn run_saga_raw_result(
+impl super::Nexus {
+    /// Lists sagas currently managed by this Nexus instance
+    pub(crate) async fn sagas_list(
         &self,
-        runnable_saga: RunnableSaga,
-    ) -> Result<SagaResult, Error> {
-        self.sec_client
-            .saga_start(runnable_saga.id)
+        opctx: &OpContext,
+        pagparams: &DataPageParams<'_, Uuid>,
+    ) -> ListResult<nexus_types::internal_api::views::Saga> {
+        // The endpoint we're serving only supports `ScanById`, which only
+        // supports an ascending scan.
+        bail_unless!(
+            pagparams.direction == dropshot::PaginationOrder::Ascending
+        );
+        opctx.authorize(authz::Action::Read, &authz::FLEET).await?;
+        let marker = pagparams.marker.map(|s| SagaId::from(*s));
+        let saga_list = self
+            .sagas
+            .sec_client
+            .saga_list(marker, pagparams.limit)
             .await
-            .context("starting saga")
-            .map_err(|error| Error::internal_error(&format!("{:#}", error)))?;
-
-        Ok(runnable_saga.fut.await)
+            .into_iter()
+            .map(nexus_types::internal_api::views::Saga::from)
+            .map(Ok);
+        Ok(futures::stream::iter(saga_list).boxed())
     }
 
-    pub fn sec(&self) -> &steno::SecClient {
-        &self.sec_client
+    /// Fetch information about a saga currently managed by this Nexus instance
+    pub(crate) async fn saga_get(
```

Comment:
Nice to see this split out; this feels essential for eventually making "async-through-the-HTTP-API" operations.